src/std_extensions/collections.rs

```diff
@@ -0,0 +1,209 @@
+//! Basic logical operations.
+
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+use smol_str::SmolStr;
+
+use crate::{
+    extension::{ExtensionSet, SignatureError, TypeDef, TypeDefBound},
+    types::{
+        type_param::{TypeArg, TypeParam},
+        CustomCheckFailure, CustomType, Type, TypeBound, TypeRow,
+    },
+    values::{CustomConst, Value},
+    Extension,
+};
+
+/// Reported unique name of the list type.
+pub const LIST_TYPENAME: SmolStr = SmolStr::new_inline("List");
+/// Pop operation name.
+pub const POP_NAME: SmolStr = SmolStr::new_inline("pop");
+/// Push operation name.
+pub const PUSH_NAME: SmolStr = SmolStr::new_inline("push");
+/// Reported unique name of the extension
+pub const RESOURCE_NAME: SmolStr = SmolStr::new_inline("Collections");
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+/// Dynamically sized list of values, all of the same type.
+pub struct ListValue(Vec<Value>);
+
+#[typetag::serde]
+impl CustomConst for ListValue {
+    fn name(&self) -> SmolStr {
+        if self.0.is_empty() {
+            "list".into()
```

Comment:
This is annoying, is it possible to add a type to `ListValue`?

Reply:
Not sure what you mean, you mean add a concrete `CustomType` as a field of
`ListValue`? How would you use that for `name()`?

Reply:
change `ListValue` to be defined like
```rust
pub struct ListValue(Vec<Value>,Type); // not sure what the right type is, substitute that for Type
```
replace this `"list".into()` line with
```rust
format!("list:<{}>",self.1.name.into())
```
of course this will then coalesce somewhat with the else branch

Reply:
ah, got it

Reply:
elsewhere in the code we've tried to keep the value-holders minimal in terms of just storing data, keeping all the type information in the type structures. I think the right move here is just to have a simple "list" here for all cases and not worry about the element type in the constant name. What do you think?

Reply:
I'm fine with your approach. 

It's not the name that's that I'm worried about (though of course that's what this code is), it's that I would expect a `CustomConst` to have a `get_type` method. But, clearly it doesn't!

Reply:
yeah we had a whole discussion about synthesisable vs. checkable types and we went for checkable across the board. #318